geps/gep-1651.md

```diff
@@ -0,0 +1,408 @@
+  //
+  // Implementations MAY support this routability
+  GatewayRoutabilityPrivate GatewayRoutability  = "Private"
+
+  // GatewayRoutabilityCluster means the Gateway's address MUST
+  // be only be routable inside the [cluster's network]
+  //
+  // Implementations MAY support this routability
+  //
+  // [cluster's network](https://kubernetes.io/docs/concepts/cluster-administration/networking/#how-to-implement-the-kubernetes-network-model)
+  GatewayRoutabilityCluster GatewayRoutability  = "Cluster"
+)
+
+type GatewaySpec struct {
+  // Infrastructure defines infrastructure level attributes about this Gateway instance.
+  Infrastructure GatewayInfrastructure `json:"infrastructure"`
+  // ...
+}
+type GatewayInfrastructure struct {
+  // Routability allows the Gateway to specify the accessibility of it's addresses. Setting
+  // this property will override the default value defined by the [GatewayClass]
+  //
+  // If the desired Gateway routability is incompatible with the [GatewayClass] implementations
+  // MUST set the condition `Accepted` to `False` with `Reason` set to `UnsupportedRoutability`.
+
+  // The default value of routability is implementation specific.
+  // It is RECOMMENDED that the default routability remains consistent for Gateways with the same
+  // gatewayClassName
+  //
+  // Implementations MAY leave this property unset and signal the default
+  // routability in the [GatewayStatus]
+  //
+  // Implementations MAY prevent end-users from updating the routability value of a Gateway.
+  // If updates are allowed the semantics and behaviour will depend on the underlying implementation.
+  // If a Gateway is mutated but does not support the desired routability it MUST set the conditions
+  // `Accepted`, `Ready`, `Programmed` to `False` with `Reason` set to `UnsupportedRoutability`.
+  //
+  // It is RECOMMENDED that in-cluster gateways SHOULD NOT support 'Private' routability.
+  // Kubernetes doesn't have a concept of 'Private' routability for Services. In the future this may
+  // change upstream.
+  //
+  // +optional
+  Routability *GatewayRoutability `json:"routability,omitempty"`
+}
+
+type GatewayStatus struct {
+	// Addresses lists the IP addresses that have actually been
+	// bound to the Gateway. These addresses may differ from the
+	// addresses in the Spec, e.g. if the Gateway automatically
+	// assigns an address from a reserved pool.
+  //
+  // Implementations that support Gateway routability MUST include an address
+  // that has the same routable semantics as defined in the Gateway spec.
+	//
+  // Implementations MAY add additional addresses in status, but they MUST be
+  // semantically less than the scope of the requested scope. For example if a
+  // user requests a `Cluster` routable Gateway then the list of addresses
+  // MUST NOT have a routability of `Public` or `Private`.
+  //
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Addresses []GatewayStatusAddress `json:"addresses,omitempty"`
```

Comment:
Indentation appears to be a mix of tabs and spaces here which results in inconsistent indentation for me.